src/document.cpp

```diff
@@ -54,199 +56,234 @@ using std::deque;
 static const char* const unsupported = "Internal error: Feature not supported in this mode.";
 static const char* const invalid_type = "$Invalid_type";
 
-string location_t::str() const
+template <typename T, typename = void>
+struct has_print_indent : std::false_type
+{};
+
+template <typename T>
+struct has_print_indent<
+    T, std::void_t<decltype(std::declval<T&>().print(std::declval<std::ostream&>, std::declval<const std::string&>))>>
+    : std::true_type
+{};
+
+template <typename T>
+constexpr auto has_print_indent_v = has_print_indent<T>::value;
+
+template <typename Item>
+std::string stringify_t<Item>::str() const
+{
+    auto os = std::ostringstream{};
+    static_cast<const Item*>(this)->print(os);
+    return os.str();
+}
+
+template <typename Item>
+std::string stringify_t<Item>::str(const std::string& indent) const
 {
-    std::string str = "LOCATION (";
-    str += uid.getName() + ", " + invariant.str() + ")";
-    return str;
+    auto os = std::ostringstream{};
+    if constexpr (has_print_indent_v<Item>)
+        static_cast<const Item*>(this)->print(os, indent);
+    else
+        static_cast<const Item*>(this)->print(os);
```

Comment:
Instead of falling back to `print(os)` I think we should be throwing a compilation error.

Because its incredibly misleading accepting an `indent` parameter and not using it for anything, so I propose we only allow `str(indent)` to be used on objects with a matching `print(os, indent)` method.

Another option is changing it to something like `static_cast<const Item*>(this)->print(os << indent);` but its not a perfect solution